paimon-core/src/main/java/org/apache/paimon/manifest/ManifestFileMeta.java

```diff
@@ -137,53 +139,68 @@ public String toString() {
-                            metas,
+                            input,
                             newMetas,
                             manifestFile,
-                            partitionType,
                             suggestedMetaSize,
-                            manifestFullCompactionSize);
-            // merge existing manifests first
-            for (ManifestFileMeta manifest : manifestFileMetas) {
-                totalSize += manifest.fileSize;
-                candidates.add(manifest);
-                if (totalSize >= suggestedMetaSize) {
-                    // reach suggested file size, perform merging and produce new file
-                    mergeCandidates(candidates, manifestFile, result, newMetas);
-                    candidates.clear();
-                    totalSize = 0;
-                }
-            }
-
-            // merge the last bit of manifests if there are too many
-            if (candidates.size() >= suggestedMinMetaCount) {
-                mergeCandidates(candidates, manifestFile, result, newMetas);
-            } else {
-                result.addAll(candidates);
-            }
+                            manifestFullCompactionSize,
+                            partitionType);
+            return fullCompacted.orElseGet(
+                    () ->
+                            tryMinorCompaction(
+                                    input,
+                                    newMetas,
+                                    manifestFile,
+                                    suggestedMetaSize,
+                                    suggestedMinMetaCount));
         } catch (Throwable e) {
             // exception occurs, clean up and rethrow
             for (ManifestFileMeta manifest : newMetas) {
                 manifestFile.delete(manifest.fileName);
             }
             throw e;
         }
+    }
 
+    private static List<ManifestFileMeta> tryMinorCompaction(
+            List<ManifestFileMeta> input,
+            List<ManifestFileMeta> newMetas,
+            ManifestFile manifestFile,
+            long suggestedMetaSize,
+            int suggestedMinMetaCount) {
+        List<ManifestFileMeta> result = new ArrayList<>();
+        List<ManifestFileMeta> candidates = new ArrayList<>();
+        long totalSize = 0;
+        // merge existing small manifest files
+        for (ManifestFileMeta manifest : input) {
+            totalSize += manifest.fileSize;
+            candidates.add(manifest);
+            if (totalSize >= suggestedMetaSize) {
+                // reach suggested file size, perform merging and produce new file
+                mergeCandidates(candidates, manifestFile, result, newMetas);
+                candidates.clear();
+                totalSize = 0;
+            }
+        }
+
+        // merge the last bit of manifests if there are too many
+        if (candidates.size() >= suggestedMinMetaCount) {
+            mergeCandidates(candidates, manifestFile, result, newMetas);
```

Comment:
Do we need to keep this logic after full compaction

Reply:
If we've done a full compaction, there shouldn't be any small files that need to be merged